File: backend/src/main/java/com/ssafy/meethub/auth/service/AuthService.java
package com.ssafy.meethub.auth.service;

import com.ssafy.meethub.auth.dto.request.ModifyPasswordRequest;
import com.ssafy.meethub.auth.dto.request.SigninRequest;
import com.ssafy.meethub.auth.dto.request.SignupRequest;
import com.ssafy.meethub.auth.dto.response.ReissuanceTokenResponse;
import com.ssafy.meethub.auth.dto.response.SigninResponse;
import com.ssafy.meethub.auth.exception.AuthErrorCode;
import com.ssafy.meethub.common.exception.BusinessException;
import com.ssafy.meethub.common.security.JwtTokenProvider;
import com.ssafy.meethub.user.entity.User;
import com.ssafy.meethub.user.repository.UserRepository;
import com.ssafy.meethub.user.service.UserService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserService userService;
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;

    @Value("${jwt.access-token-validity}")
    private long accessTokenValidity;

    @Transactional
    public void signup(SignupRequest request) {
        userService.createUser(request.email(), request.name(), request.password());
    }

    @Transactional(readOnly = true)
    public SigninResponse signin(SigninRequest request) {
        User user = userRepository.findByEmailAndIsDeletedFalse(request.email())
                .orElseThrow(() -> new BusinessException(AuthErrorCode.INVALID_CREDENTIALS));

        if (user.isDeleted()) {
            throw new BusinessException(AuthErrorCode.INVALID_CREDENTIALS);
        }

        if (!passwordEncoder.matches(request.password(), user.getPassword())) {
            throw new BusinessException(AuthErrorCode.INVALID_CREDENTIALS);
        }

        String accessToken = jwtTokenProvider.createAccessToken(user.getId());
        String refreshToken = jwtTokenProvider.createRefreshToken(user.getId());

        return new SigninResponse(user.getId(), accessToken, "Bearer", accessTokenValidity / 1000);
    }

    public Long getUserIdFromAccessToken(String accessToken) {
        return jwtTokenProvider.getUserId(accessToken);
    }

    public String getRefreshToken(Long userId) {
        return jwtTokenProvider.getRefreshToken(userId);
    }

    @Transactional
    public void signout(Long userId, String accessToken) {
        // Access Token을 블랙리스트에 추가
        if (accessToken != null && !accessToken.isEmpty()) {
            jwtTokenProvider.addToBlacklist(accessToken);
        }

        // userId로 직접 Refresh Token 삭제
        jwtTokenProvider.deleteRefreshTokenByUserId(userId);
    }

    @Transactional
    public ReissuanceTokenResponse reissuanceToken(String refreshToken, String oldAccessToken) {
        if (!jwtTokenProvider.validateRefreshToken(refreshToken)) {
            throw new BusinessException(AuthErrorCode.INVALID_TOKEN);
        }

        Long userId = jwtTokenProvider.getUserId(refreshToken);

        if (oldAccessToken != null && !oldAccessToken.isEmpty()) {
            jwtTokenProvider.addToBlacklist(oldAccessToken);
        }

        jwtTokenProvider.deleteRefreshTokenByUserId(userId);

        String newAccessToken = jwtTokenProvider.createAccessToken(userId);
        String newRefreshToken = jwtTokenProvider.createRefreshToken(userId);

        return new ReissuanceTokenResponse(newAccessToken, "Bearer", accessTokenValidity / 1000);
    }

    @Transactional
    public void modifyPassword(Long userId, ModifyPasswordRequest request) {
        userService.modifyPassword(userId, request.currentPassword(), request.newPassword());
    }

    @Transactional(readOnly = true)
    public Boolean isDuplicatedEmail(String email) {
        return userService.isDuplicatedEmail(email);
    }
}

File: frontend/src/api/authService.ts
import apiClient from './axiosInstance';

import type { ApiResponse } from '../types/api';
import type {
  SignUpRequest,
  SignInRequest,
  SignInResponse,
  ReissuanceTokenResponse,
  UserProfile
} from '../types/auth';

// 회원가입 API
export const signUp = async (
  signUpData: SignUpRequest
): Promise<ApiResponse<null>> => {
  const response = await apiClient.post<ApiResponse<null>>(
    '/auth/signup',
    signUpData
  );
  return response.data;
};

// 로그인 API
export const signIn = async (
  signInData: SignInRequest
): Promise<ApiResponse<SignInResponse>> => {
  const response = await apiClient.post<ApiResponse<SignInResponse>>(
    '/auth/signin',
    signInData
  );
  return response.data;
};

// 사용자 프로필 조회 API
export const getUserProfile = async (): Promise<UserProfile> => {
  const response = await apiClient.get<ApiResponse<UserProfile>>('/me');
  return response.data.content; 
};

// 프로필 수정 API
export const profileEdit = async (name: string): Promise<ApiResponse<null>> => {
  const response = await apiClient.patch<ApiResponse<null>>('/me', { name });
  return response.data;
};

// 탈퇴 API
export const withdraw = async (): Promise<ApiResponse<null>> => {
  const response = await apiClient.delete<ApiResponse<null>>('/me');
  return response.data;
};

// 로그아웃 API
export const signOut = async (): Promise<ApiResponse<null>> => {
  const response = await apiClient.post<ApiResponse<null>>('/auth/signout');
  return response.data;
};

// 이메일 중복 확인 API
export const checkEmailDuplicate = async (
  email: string
): Promise<ApiResponse<boolean>> => {
  const response = await apiClient.post<ApiResponse<boolean>>(
    '/auth/check-email',
    { email }
  );
  return response.data;
};

// 토큰 재발급 API
export const refreshAccessToken = async (): Promise<ApiResponse<ReissuanceTokenResponse>> => {
  const response = await apiClient.post<ApiResponse<ReissuanceTokenResponse>>('/auth/refresh');
  return response.data;
};

// 비밀번호 변경 API
export const changePassword = async (
  currentPassword: string,
  newPassword: string
): Promise<ApiResponse<null>> => {
  const response = await apiClient.post<ApiResponse<null>>('/auth/password', {
    current_password: currentPassword,
    new_password: newPassword
  });
  return response.data;
};

